cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll

```diff
@@ -651,66 +897,161 @@ class BinaryInstruction extends Instruction {
+ *
+ * The operand must have a numeric type, which will also be the result type. The result of integer
+ * negation uses two's complement, and is computed modulo 2^n. The result of floating-point negation
+ * is performed according to IEEE-754.
+ */
 class NegateInstruction extends UnaryArithmeticInstruction {
   NegateInstruction() { getOpcode() instanceof Opcode::Negate }
 }
 
+/**
+ * An instruction that computes the result of a bitwise operation.
+ */
 class BitwiseInstruction extends Instruction {
   BitwiseInstruction() { getOpcode() instanceof BitwiseOpcode }
 }
 
+/**
+ * An instruction that performs a bitwise operation on two integer operands.
+ */
 class BinaryBitwiseInstruction extends BitwiseInstruction, BinaryInstruction { }
 
+/**
+ * An instruction that performs a bitwise operation on a single integer operand.
+ */
 class UnaryBitwiseInstruction extends BitwiseInstruction, UnaryInstruction { }
 
+/**
+ * An instruction that computes the bitwise "and" of two integer operands.
+ *
+ * Both operands must have the same integer type, which will also be the result type.
+ */
 class BitAndInstruction extends BinaryBitwiseInstruction {
   BitAndInstruction() { getOpcode() instanceof Opcode::BitAnd }
 }
 
+/**
+ * An instruction that computes the bitwise "or" of two integer operands.
+ *
+ * Both operands must have the same integer type, which will also be the result type.
+ */
 class BitOrInstruction extends BinaryBitwiseInstruction {
   BitOrInstruction() { getOpcode() instanceof Opcode::BitOr }
 }
 
+/**
+ * An instruction that computes the bitwise "xor" of two integer operands.
+ *
+ * Both operands must have the same integer type, which will also be the result type.
+ */
 class BitXorInstruction extends BinaryBitwiseInstruction {
   BitXorInstruction() { getOpcode() instanceof Opcode::BitXor }
 }
 
+/**
+ * An instruction that shifts its left operand to the left by the number of bits specified by its
+ * right operand.
+ *
+ * Both operands must have an integer type. The result has the same type as the left operand. The
+ * rightmost bits are zero-filled.
+ */
 class ShiftLeftInstruction extends BinaryBitwiseInstruction {
   ShiftLeftInstruction() { getOpcode() instanceof Opcode::ShiftLeft }
 }
 
+/**
+ * An instruction that shifts its left operand to the right by the number of bits specified by its
+ * right operand.
+ *
+ * Both operands must have an integer type. The result has the same type as the left operand. If the
+ * left operand has an unsigned integer type, the leftmost bits are zero-filled. If the left operand
+ * has a signed integer type, the leftmost bits are filled by duplicating the most significant bit
+ * of the left operand.
+ */
 class ShiftRightInstruction extends BinaryBitwiseInstruction {
   ShiftRightInstruction() { getOpcode() instanceof Opcode::ShiftRight }
 }
 
+/**
+ * An instruction that performs a binary arithmetic operation involving at least one pointer
+ * operand.
```

Comment:
If there's more than one pointer, do they have to have the same type?  (I think this is more-or-less implied by the existence of `getElementSize`)

Reply:
Yes. I've now clarified that in the comment for `PointerDiffInstruction`, which is the only one with two pointer operands. Note that in the IR type system, pointers do not have referent types, so "the same pointer type" means "both are pointer types with the same size", which is typically all pointer types. So, technically the IR can represent subtracting a `void*` from a `const volatile double*`, although none of the languages are capable of generating such a construct.